backend/src/Parser/FSharpToProgramTypes.fs

```diff
@@ -0,0 +1,762 @@
+/// Converts F# types to Dark ProgramTypes
+module FSharpToProgramTypes
+
+open FSharp.Compiler.Syntax
+
+open Prelude
+open Tablecloth
+
+module PT = LibExecution.ProgramTypes
+
+
+// A placeholder is used to indicate what still needs to be filled
+let private placeholder =
+  PT.EString(12345678UL, [ PT.StringText "PLACEHOLDER VALUE" ])
+
+// This is a "Partial active pattern" that you can use as a Pattern to match a Placeholder value
+let private (|Placeholder|_|) (input : PT.Expr) =
+  if input = placeholder then Some() else None
+
+// TODO: support custom types that aren't the 0th version of the type
+let private matchingCustomTypes
+  // TODO: we don't used the defs here - just the names
+  (availableTypes : List<PT.FQTypeName.T * PT.CustomType.T>)
+  (nameOfType : string)
+  : List<PT.FQTypeName.T> =
+  availableTypes
+  |> List.choose (fun (typeName, _def) ->
+    match typeName with
+    | PT.FQTypeName.User u -> if u.typ = nameOfType then Some typeName else None
+    | PT.FQTypeName.Stdlib t -> if t.typ = nameOfType then Some typeName else None)
+
+
+module DType =
+  let rec fromSynType
+    (availableTypes : List<PT.FQTypeName.T * PT.CustomType.T>)
+    (typ : SynType)
+    : PT.DType =
+
+
+    let c = fromSynType availableTypes
+
+    let rec fromNameAndTypeArgs
+      availableTypes
+      (name : string)
+      (typeArgs : List<SynType>)
+      : PT.DType =
+      match name, typeArgs with
+      // no type args
+      | "bool", [] -> PT.TBool
+      | "int", [] -> PT.TInt
+      | "string", [] -> PT.TStr
+      | "char", [] -> PT.TChar
+      | "float", [] -> PT.TFloat
+      | "DateTime", [] -> PT.TDateTime
+      | "UUID", [] -> PT.TUuid
+      | "unit", [] -> PT.TUnit
+      | "Password", [] -> PT.TPassword
+
+      // with type args
+      | "List", [ arg ] -> PT.TList(c arg)
+      | "Option", [ arg ] -> PT.TOption(c arg)
+
+      // Some user- or stdlib- type
+      | _ ->
+        match matchingCustomTypes availableTypes name with
+        | [ matchedType ] -> PT.TCustomType(matchedType, List.map c typeArgs)
+        | _ ->
+          Exception.raiseInternal $"Unsupported type" [ "name", name; "type", typ ]
```

Comment:
TODO: note in the err msg that we didn't find exactly 1 matching type (could be 0, or 2+)